### PR TITLE
CDAP-1897: Show config entry source in JSON for ConfigHandler

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-1897
http://builds.cask.co/browse/CDAP-DUT1295

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
There are no tests. Can you add a test?

**Comment by nitinmotgi** [comment]
@jwang47 can you address this immediately.

**Comment by nitinmotgi** [comment]
:-1: There is a major flaw with this approach. Router might be outside of firewall and might not have the configurations available for it. So, these endpoints should be dispatched to master. Please address this issue immediately.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 import co.cask.cdap.common.utils.DirUtils;
 import com.google.common.base.Preconditions;
 import com.google.gson.stream.JsonWriter;
+import org.apache.hadoop.classification.InterfaceStability;
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java]: Why use hadoop annotation in common package?

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java]: This is an internal class, ok not to have annotation like this. If needed, use `@Beta`.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java [hunk 2]
```diff
@@ -997,6 +998,44 @@ public void setPattern(String name, Pattern pattern) {
...
+      // If properties is null, it means a resource was newly added
+      // but the props were cleared so as to load it upon future
+      // requests. So lets force a load by asking a properties list.
+      getProps();
+    }
+    // Return a null right away if our properties still
+    // haven't loaded or the resource mapping isn't defined
+    if (properties == null || updatingResource == null) {
+      return null;
+    } else {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java]: No need to have else after return statement in the if block.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java [hunk 3]
```diff
@@ -997,6 +998,44 @@ public void setPattern(String name, Pattern pattern) {
...
+      String resource = updatingResource.get(name);
+      if (resource == null) {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java]: Use `return (resources == null) ? null : new String[] { resource };`

File: cdap-common/src/main/java/co/cask/cdap/gateway/handlers/ConfigHandler.java [hunk 4]
```diff
@@ -97,15 +98,60 @@ public void configHBase(@SuppressWarnings("UnusedParameters") HttpRequest reques
...
-    ImmutableMap.Builder<String, String> result = ImmutableMap.builder();
+  private String getFirstElement(String[] array) {
+    if (array != null && array.length >= 1) {
+      return array[0];
+    } else {
+      return null;
+    }
+  }
+
+  private List<ConfigEntry> toObject(Configuration configuration) {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/gateway/handlers/ConfigHandler.java]: Rather than super generic name `toObject`, better call it `toConfigEntries` or `toConfigEntryList`.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java (hunk 2 above)
Comment (jwang47) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java]: This method was copied from Hadoop's Configuration class. Should I just add a note without refactoring it?

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java [hunk 5]
```diff
@@ -997,6 +998,43 @@ public void setPattern(String name, Pattern pattern) {
     }
   }
 
+
+  /**
+   * From org.apache.hadoop.conf.Configuration, with minor modifications.
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java]: No need to mention this.

**Comment by jwang47** [comment]
Updated PR to run `ConfigHandler` from master.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/CConfiguration.java [hunk 6]
```diff
@@ -34,7 +34,7 @@
   private static final Logger LOG =
     LoggerFactory.getLogger(CConfiguration.class);
 
-  private CConfiguration() {
+  public CConfiguration() {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/CConfiguration.java]: why the change? And it's different than what the comment inside this constructor.

File: cdap-common/src/test/java/co/cask/cdap/gateway/handlers/ConfigServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ *
+ */
+public class ConfigServiceTest {
+
+  @Test
+  public void testConfig() {
+
+    // cConf
+    CConfiguration cConf = new CConfiguration();
```
Comment (chtyim) [inline cdap-common/src/test/java/co/cask/cdap/gateway/handlers/ConfigServiceTest.java]: Shouldn't do this. You can do a `cConf.clear()`.

File: cdap-common/src/main/java/co/cask/cdap/gateway/handlers/ConfigHandler.java [hunk 8]
```diff
@@ -62,12 +55,11 @@ public ConfigHandler(CConfiguration cConf, Configuration hConf) {
   public void configCDAP(@SuppressWarnings("UnusedParameters") HttpRequest request, HttpResponder responder,
                         @DefaultValue("json") @QueryParam("format") String format) {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/gateway/handlers/ConfigHandler.java]: mis-align.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java [hunk 9]
```diff
@@ -997,6 +998,41 @@ public void setPattern(String name, Pattern pattern) {
...
+   * @return null - If the property or its source wasn't found. Otherwise,
+   * returns a list of the sources of the resource.  The older sources are
+   * the first ones in the list.  So for example if a configuration is set from
+   * the command line, and then written out to a file that is read back in the
+   * first entry would indicate that it was set from the command line, while
+   * the second one would indicate the file that the new configuration was read
+   * in from.
+   */
+  @Beta
+  public synchronized String[] getPropertySources(String name) {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java]: If it will return null, add `@Nullable`. BTW, why not return an empty array if property not found?

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java [hunk 10]
```diff
@@ -186,6 +186,11 @@ private String getV3RoutingService(String [] uriParts, AllowedMethod requestMeth
       // /v3/namespaces/{namespace-id}/data/datasets/{name}/properties
       // /v3/namespaces/{namespace-id}/data/datasets/{name}/admin/{method}
       return Constants.Service.DATASET_MANAGER;
+    } else if (uriParts.length >= 2 && uriParts[1].equals("config")) {
+      // TODO: config endpoint only hits stream service
```
Comment (chtyim) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: What does this comment mean? config endpoint hits stream service??

Comment (jwang47) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: Router will route calls to /v3/config/* to stream service. Wasn't sure where would be a better place to route /v3/config/* requests and run `ConfigHandler`, so right now it's hitting stream service.

Comment (chtyim) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: You mean the stream handler? So the stream handler is handling /v3/config calls to give the cConf?

Comment (jwang47) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: It's the `StreamHttpService` that runs the `ConfigHandler` -- see https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamServiceModule.java#L39: `CommonHandlers.add(handlerBinder);` adds the `ConfigHandler`.

Comment (chtyim) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: So we are having the config handler in every http service? That seems very wrong. Why not route to master (app-fabric)?

Comment (jwang47) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: True, need to route to app-fabric.

**Comment by sreevatsanraman** [comment]
LGTM :+1:
